coordinape/coordinape
# Implement AdminUpdateUser Hasura Endpoint

In addition to implementing the endpoint I've made the following
changes:

1. Refactored optional profile creation (as a part of other mutations)
   into an event trigger based on user creation or updates
2. Created composable discord and telegram factories that work with
   logic handlers for event triggers that emit bot messages, based on
   @jsur's existing work. I'm considering combining the factories into
   a single function and then allowing the triggers to emit the variable
   that determines whether or not discord or telegram is the target.
3. Extended @jsur's event trigger payload type into a dynamic type
   that changes based on the table and type of operation on the table
   that emits the event.

# Test Plan

1. Follow the README instructions to set up the environment
2. find the next epoch in the database, if one isn't already active
and adjust the start date to before the current date. Take note of the
circle
3. Allocate some give to members of the Circle in the UI.
4. Set up a discord webhook for the circle and add a telegram channel id
in the circle's database entry
5. pick a user in the circle give has been allocated to and execute the
following mutation in the graphiql interface:
```gql
mutation UpdateUserTest {
  adminUpdateUser(
    object: {
      address: "<recipient's address>",
      circle_id: <recipient circleID>,
      non_receiver: true
    }
  ) {
    address
  }
}
```

6. verify that discord and telegram entries are created for the opt-out
   and the refund. The total refunded from the opted-out user will
   likely be incorrect due to the mock data in the database, but refund
   to the user you allocated from should be correct.
7. Notice that a new profile was created with a lower-case version of
   the user's address, since the existing address is checksum-mixed
   case. This can be verified in the create-profile event trigger's
   "processed events" tab.
8. run other queries and verify that fixed_non_receiver and non_receiver
   behave correctly as inputs together, along with give_token_remaining.

# TODO
- [x] run a couple more tests and ensure error responses are sensible
- [x] implement telegram and discord bot hooks.
- [x] convert profile check and creation into an event trigger
- [x] configure and test event triggers in hasura
- [x] Seems like @jsur has made some progress on bot comms in hasura so I'll try to borrow from their work.

resolves https://github.com/coordinape/coordinape/issues/518
